spinoso-string/src/lib.rs

```diff
@@ -1682,27 +1670,22 @@ impl String {
     /// use spinoso_string::String;
     ///
     /// let s = String::utf8("via 💎 v3.2.0".as_bytes().to_vec());
-    /// assert_eq!(s.rindex("v", None), Some(9)); // FIXME: Should be 5 (#2360)
+    /// assert_eq!(s.rindex("v", None), Some(6));
+    /// assert_eq!(s.rindex("v", Some(5)), Some(0));
+    /// assert_eq!(s.rindex("v", Some(6)), Some(6));
     /// assert_eq!(s.rindex("a", None), Some(2));
     /// ```
     ///
     /// [`String#rindex`]: https://ruby-doc.org/core-3.1.2/String.html#method-i-rindex
     #[inline]
     #[must_use]
     pub fn rindex<T: AsRef<[u8]>>(&self, needle: T, offset: Option<usize>) -> Option<usize> {
-        fn inner(buf: &[u8], needle: &[u8], offset: Option<usize>) -> Option<usize> {
-            if let Some(offset) = offset {
-                let end = buf.len().checked_sub(offset).unwrap_or_default();
-                let buf = buf.get(..end)?;
-                buf.rfind(needle)
-            } else {
-                buf.rfind(needle)
-            }
+        let default_max_offset = self.inner.char_len() - 1;
+        let mut offset = offset.unwrap_or(default_max_offset);
```

Comment:
`char_len()` is potentially an expensive operation since it must do a full (UTF-8) decode of the string. Please move it to be lazily evaluated in:

```suggestion
        let mut offset = offset.unwrap_or_else(|| self.inner.char_len().checked_sub(1).unwrap_or_default());
```

I would have expected this underflow on an empty string to have resulted in a panic in tests. Can we add some tests for both empty receivers and empty needles?

Reply:
Thanks for pointing out. The underflow indeed can be reproduced with 1847a6a1b2e15dd171b75b2413bdebd179e8b576.

Fixed in 4cd5c98add24b9945b28627b77c69c05ff174250.